rootston/cursor.c

```diff
@@ -30,19 +30,81 @@ void roots_cursor_destroy(struct roots_cursor *cursor) {
 	// TODO
 }
 
+static void seat_view_deco_motion(struct roots_seat_view *view, double deco_vx, double deco_vy) {
+	struct roots_cursor *cursor = view->seat->cursor;
+
+	double vx = deco_vx;
+	double vy = deco_vy;
+	if (view->has_button_grab) {
+		vx = view->grab_vx;
+		vy = view->grab_vy;
+	}
+
+	bool is_titlebar = vy < 0 && -vy < view->view->titlebar_height;
+	uint32_t edges = 0;
+	if (vx < 0) {
+		edges |= WLR_EDGE_LEFT;
+	} else if (vx > view->view->wlr_surface->current->width) {
+		edges |= WLR_EDGE_RIGHT;
+	} else if (vy > view->view->wlr_surface->current->height) {
+		edges |= WLR_EDGE_BOTTOM;
+	} else if (-vy > view->view->titlebar_height) {
+		edges |= WLR_EDGE_TOP;
+	}
+
+	if (view->has_button_grab) {
+		if (is_titlebar) {
+			roots_seat_begin_move(view->seat, view->view);
+		} else if (edges) {
+			roots_seat_begin_resize(view->seat, view->view, edges);
+		}
+		view->has_button_grab = false;
+	} else {
+		if (is_titlebar) {
+			wlr_xcursor_manager_set_cursor_image(cursor->xcursor_manager,
+				cursor->default_xcursor, cursor->cursor);
+		} else if (edges) {
+			const char *resize_name = wlr_xcursor_get_resize_name(edges);
+			wlr_xcursor_manager_set_cursor_image(cursor->xcursor_manager,
+				resize_name, cursor->cursor);
+		}
+	}
+}
+
+static void seat_view_deco_leave(struct roots_seat_view *view) {
+	view->has_button_grab = false;
+}
+
+static void seat_view_deco_button(struct roots_seat_view *view, double vx,
+		double vy, uint32_t button, uint32_t state) {
+	if (button == BTN_LEFT && state == WLR_BUTTON_PRESSED) {
+		view->has_button_grab = true;
+		view->grab_vx = vx;
+		view->grab_vy = vy;
+	} else {
+		view->has_button_grab = false;
+	}
+}
```

Comment:
I would prefer to establish a decorations file and move this (and other relevant logic) there.